feat(reaction_analyzer): add reaction anaylzer tool to measure end-to-end delay in sudden obstacle braking response

brkay54 opened the pull request:
## Description

Main issue:
- https://github.com/autowarefoundation/autoware.universe/issues/6547

Strictly depends on:
- https://github.com/autowarefoundation/autoware.universe/pull/6382

Optionally depends on:
- https://github.com/autowarefoundation/autoware.universe/issues/6255

The main purpose of the reaction analyzer package is to measure the reaction times of the nodes by listening to the
pre-determined topics. It can be used for measuring the reaction time of the perception pipeline, planning pipeline, and
control pipeline. To be able to measure both control outputs and perception outputs, it is necessary to divide the node
into two running_mode: `planning_control` and `perception_planning`.

### Planning Control Mode

In this mode, the reaction analyzer creates a dummy publisher for the PredictedObjects and PointCloud topics. In the
beginning of the test, it publishes the initial position of the ego vehicle and the goal position to set the test
environment. Then, it spawns a sudden obstacle in front of the ego vehicle. After the obstacle is spawned, it starts to
measure the pre-determined topics to catch the reacted messages of the planning and control nodes. When all the topics are reacted, it calculates the reaction time of the nodes and statistics, and it creates a csv file to store the
results.

### Perception Planning Mode

In this mode, the reaction analyzer reads the rosbag files which are recorded from AWSIM, and it creates a topic
publisher for each topic to replay the rosbag. It reads two rosbag files: `path_bag_without_object`
and `path_bag_with_object`. Firstly, it replays the `path_bag_without_object` to set the initial position of the ego
vehicle and the goal position. After `spawn_time_after_init` seconds, it replays the `path_bag_with_object` to spawn a
sudden obstacle in front of the ego vehicle. After the obstacle is spawned, it starts to measure the pre-determined
topics to catch the reacted messages of the perception and planning nodes. When all the topics are reacted, it
calculates
the reaction time of the nodes and statistics, and it creates a CSV file to store the results.

<!-- Write a brief description of this PR. -->

## Related links
Related Issue:
- https://github.com/autowarefoundation/autoware.universe/issues/5540

## Tests performed
You can see the results in the Excel table [here](https://docs.google.com/spreadsheets/d/1qkSAMAeYa1taIg3HsUJKYvRS2nIYbMWsq9-b7bN5Dfw/edit?usp=sharing). You can see the test dates below the page.
Also, some tests were shared in issues comments: https://github.com/autowarefoundation/autoware.universe/issues/5540
<!-- Describe how you have tested this PR. -->

## Notes for reviewers

This PR currently depends on other PR below:
- https://github.com/autowarefoundation/autoware.universe/pull/6382


<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Interface changes

<!-- Describe any changed interfaces, such as topics, services, or parameters. -->

## Effects on system behavior

<!-- Describe how this PR affects the system behavior. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

xmfcx commented on tools/reaction_analyzer/src/reaction_analyzer_node.cpp:
```diff
@@ -0,0 +1,403 @@
+}
+
+// ---
+void ReactionAnalyzerNode::pointcloudCallback(const PointCloud2::ConstSharedPtr msg_ptr)
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+  const auto seen_time = this->now();
+  // check pointcloud spawned or not
+  if (msg_ptr->data.empty()) return;
+  if (!pointcloud_seen_time_) {
+    pointcloud_seen_time_ = seen_time;
+  }
+}
```
This should be the message stamp right?

brkay54 commented:
** Weekly Update **

* Last week, to be able to run the perception pipeline reaction test, we ran the reaction_analyzer by playing rosbag in another terminal. This causes ~10ms time resolution while measuring the reaction times. This week, I made some changes to the reaction analyzer to be able to record the necessary messages of the rosbag and replay them inside the node. In this way, we can measure the reaction times in the system time.

* 3 Lidar test environment created by using AWSIM, and I can run perception pipeline tests by using them successfully. 

* Firstly, I investigated the pointcloud_preprocess pipeline to analyze how much delay we got to get concatenated pointcloud output. For this test, I used 3 Lidar setup and I analyzed the `pipeline_latency_ms` which we added [before](https://github.com/autowarefoundation/autoware.universe/pull/6056):

When I published 3 lidar pointcloud outputs, the pipeline_latency_ms started with a lower latency value:
 
![Screenshot from 2024-02-05 06-11-03](https://github.com/autowarefoundation/autoware.universe/assets/45468306/24f9abbb-b1be-4318-b38f-e45f3f0572c0)

After 12 minutes:

![Screenshot from 2024-02-05 06-23-43](https://github.com/autowarefoundation/autoware.universe/assets/45468306/35ae9f05-f27d-4257-b16d-762bc75d8fb6)

As you can see, the latency value of the concatenated nodelet of the right lidar is accumulating.

After 10 minutes: 

![Screenshot from 2024-02-05 06-32-52](https://github.com/autowarefoundation/autoware.universe/assets/45468306/c50d3851-b9b5-4332-b718-e6339107fb6e)

After this higher value, the concatenated pointcloud delay decreases immediately as you can see graph below:

![Screenshot from 2024-02-05 06-33-46](https://github.com/autowarefoundation/autoware.universe/assets/45468306/ef6d6d5f-6512-4aba-8b44-f9527d7f2988)

It seems to me a weird behavior. After, some time, I thought that it would be caused by phase differences between lidar output publishers. Then, I added another feature to the reaction analyzer to make it able to publish pointcloud outputs synchronically.

After running the reaction analyzer with synchronically published pointcloud, I couldn't see this accumulated pipeline_latency on the concatenating process:
 
![Screenshot from 2024-02-05 06-57-55](https://github.com/autowarefoundation/autoware.universe/assets/45468306/b4061c51-5c60-4a8b-8de3-686dac8ea7c1)

The concatenated pointcloud delay of the top lidar is higher than others. It is caused by the higher delay of the distortion corrector. Distortion_corrector in the top lidar process takes more time than others because the pointcloud size of the crop_box_filtered_mirror of the top lidar is higher than others. The difference in width between the top lidar and right-left lidars is ~13,000 and it causes ~10ms delay. However, because of this ~10ms delay, the concatenate process is always running with the past pointcloud output of the top lidar. (I didn't try yet but I saw that we can set offset values for the specific pointclouds w.r.t documentation of the concatenate nodelet. I am going to try it.)

----

* For the development phase of the reaction_analyzer, all features were added, just need to clean up the code. And we need to decide how to implement the timestamp reporting process of the perception pipeline. I created an issue for it [here](https://github.com/autowarefoundation/autoware.universe/issues/6255).

yukkysaito commented on perception/tracking_object_merger/src/decorative_tracker_merger.cpp:
```diff
@@ -250,9 +250,9 @@ bool DecorativeTrackerMergerNode::decorativeMerger(
-  if (input_objects_msg->objects.empty()) {
-    return false;
-  }
+  //  if (input_objects_msg->objects.empty()) {
+  //    return false;
+  //  }
```
May I ask why this code is not required?
cc @YoshiRi

YoshiRi replied:
With this statement, tracking object merger can not update tracklet with empty input. See https://github.com/autowarefoundation/autoware.universe/issues/6373.
It already fixed, so this part is no longer needed.

yukkysaito commented on perception/map_based_prediction/src/map_based_prediction_node.cpp:
```diff
@@ -1112,7 +1112,7 @@ void MapBasedPredictionNode::objectsCallback(const TrackedObjects::ConstSharedPt
     }
   }
-
+  output.header.stamp = this->now();
```
The timestamp in processed data should fundamentally represent the time when the data was originally captured, not the time when it was processed.

For example, when performing object detection using images from a camera, it would be better to use the camera's capture time as the timestamp. This can be achieved by inheriting the timestamp from the original input. However, there are exceptions. For instance, when using a Kalman filter or similar techniques to provide delay guarantees, it might be acceptable to adjust the timestamp to the estimated time.

brkay54 replied:
@yukkysaito Thanks for the review, actually these are the temporary commits I mentioned above. You are right. I created an issue for this problem. We can not detect the publishing time in pipelines Autoware. So, we are thinking about a new method to report the message publishing times inside the nodes: https://github.com/autowarefoundation/autoware.universe/issues/6255

yukkysaito replied:
@brkay54 Thanks for the replies! If you're still in the process of sorting it out, I'll change to draft PR.
If you're at the point where you're ready to review it, please revert back to the original.

mitsudome-r commented on tools/reaction_analyzer/README.md:
```diff
@@ -0,0 +1,119 @@
+pre-determined topics. It can be used for measuring the reaction time of the perception pipeline, planning pipeline, and
+control pipeline. To be able to measure both control outputs and perception outputs, it is necessary to divide the node
+into two running_mode: `planning_control` and `perception_planning`.
+
```
Maybe you can add some images of node diagram to show how the analyzer calculates the reaction time.

brkay54 commented:
cc @mitsudome-r @xmfcx 
The blockers are ready for review. I added a class utility to enable or disable easily the published_time:

by order of merge
- https://github.com/autowarefoundation/autoware_msgs/pull/83
- https://github.com/autowarefoundation/autoware.universe/pull/6440
- https://github.com/autowarefoundation/autoware.universe/pull/6490

brkay54 commented:
Sorry for the delay, I need to convert the PR because of the implementation of the message_filters package.

The package is ready for review now. I set up the `message_filters` package into reaction_analyzer and cleaned up the code. 

This PR currently depends on other PRs below:
- https://github.com/autowarefoundation/autoware.universe/pull/6382
- https://github.com/autowarefoundation/autoware.universe/pull/6440
- https://github.com/autowarefoundation/autoware.universe/pull/6490 (not strictly)

xmfcx commented:
@brkay54 could you fix the errors stated in the pre-commit.ci?

brkay54 commented:
@xmfcx Thank you, I fixed.

xmfcx commented on tools/reaction_analyzer/include/subscriber.hpp:
```diff
@@ -0,0 +1,250 @@
+  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
+
+  // Functions
+  void initReactionChainsAndParams();
```
```suggestion
  void initReactionChainsAndParams();
```

Could you refactor all function names to snake_case?

xmfcx commented:
/review

xmfcx commented:
/improve --extended

brkay54 commented:
/review

brkay54 commented:
@xmfcx I implemented some of the automated-review suggestions, cleaned-up the code, and changed `.csv` output:
**Before:**
![Screenshot from 2024-03-18 13-32-49](https://github.com/autowarefoundation/autoware.universe/assets/45468306/5c9d6b10-6552-4f6c-aad2-8632c869d1f0)

**After:**
![Screenshot from 2024-03-18 13-33-16](https://github.com/autowarefoundation/autoware.universe/assets/45468306/87fb5e7f-031c-438b-a20e-82d83151d679)
- _NL means node latency, TL means total latency_

With the last commit, we can see each pipeline and its latencies by using the header time of the `PublishedTime.msg`. By using this, we can see if there are dropped messages or not while test runng.

Could you check the PR when you are available?

xmfcx commented:
@brkay54 could you fix the clang-tidy errors?
- https://github.com/autowarefoundation/autoware.universe/actions/runs/8326653509/job/22782747570?pr=5954#step:8:7763

brkay54 commented:
@xmfcx Clang-tidy issues were fixed ⭐

mitsudome-r commented:
I will test this PR against the main branch and merge it.

brkay54 commented on tools/reaction_analyzer/README.md:
```diff
@@ -0,0 +1,202 @@
+  and `path_bag_with_object`.
+- Because custom sensor setup, you need to check out the following branches before launch the
+  reaction analyzer: For the `autoware_individual_params` repository, check out the
+  branch [here](https://github.com/brkay54/autoware_individual_params/tree/bk/reaction-analyzer-config).
```
Yes you are right @mitsudome-r-san. Actually, main reason of the these branches is we need multiple lidar setup to test sensing pipeline accurately (to run concat and time sync process), however, current awsim sensor configuration is not supporting the multiple sensor currently. It might be planned to add multiple lidar in AWSIM sensor_kit. If it will be added, we can record a new rosbag to run the `reaction_analyzer` and we won't need to switch the branches in `awsim_sensor_kit_launch` and `autoware_individual_params`. I guess there would be some plans to add multiple lidar sensor setup to AWSIM. 

@xmfcx If there is a plan to add multiple lidar to AWSIM in near future, we can use branches in my remote (for awsim_sensor_kit_launch and autoware_individual_params) temporarily. After AWSIM sensor_kit updated, we can update the reaction_analyzer to be able to run with updated awsim_sensor_kit by default. 

What do you thing about this plan?

mitsudome-r replied:
@brkay54 When you were testing in your local environment, were you testing with customized AWSIM with multiple lidars? I think the default AWSIM binary doesn't have multiple LIDARs attached.

brkay54 replied:
> @brkay54 When you were testing in your local environment, were you testing with customized AWSIM with multiple lidars?

@mitsudome-r Yes, I was using customized AWSIM setup.

>  I think the default AWSIM binary doesn't have multiple LIDARs attached.

Yes, currently the setup is with 1 Lidar. Just I was assuming multiple lidar setup would be added to AWSIM in near future (maybe @xmfcx inform us).
My suggestion was lets merge this package with this AWSIM customization, after the AWSIM Setup is updated with multiple lidar setup, I can update the usage part of the readme with respect to latest update in AWSIM.

mitsudome-r replied:
remaining action items:
- create a pull request against https://github.com/RobotecAI/awsim_sensor_kit_launch/pulls and https://github.com/autowarefoundation/autoware_individual_params/pulls for multi-lidar support.
- Test it with AWF AWSIM main branch

brkay54 replied:
I run the some updated version of AWSIM with 3 - Lidar setup,  I couldn't see any issue. 

- https://github.com/autowarefoundation/AWSIM/pull/54#pullrequestreview-2008680579

After the PR merged, I am going to record the rosbags inside AWSIM to run the reaction_analyzer.

brkay54 replied:
I didn't realize the non-ground points on the road, it was my fault. I asked my colleague to help us to fix the transformations for sensors. 

https://github.com/autowarefoundation/autoware.universe/assets/45468306/791a0c65-519c-4468-aa2b-5ebfab171239

mitsudome-r replied:
@mozhoku Please help Berkay to check if the sensor calibration data in https://github.com/autowarefoundation/awsim_labs_sensor_kit_launch/tree/main/awsim_labs_sensor_kit_description/config match with the values in Unity.

mozhoku replied:
@brkay54 I've updated the right VLP16 params here:
- https://github.com/autowarefoundation/autoware_individual_params/pull/48

brkay54 replied:
cc @xmfcx @mitsudome-r, I updated the package w.r.t. new awsim_labs configuration. I tested both planning_control and perception_planning modes, I couldn't see any problem.

Current Scene Change for perception_planning mode:

![cap-2024-05-02-12-32-35-ezgif com-video-to-gif-converter](https://github.com/autowarefoundation/autoware.universe/assets/45468306/0603a24a-f0d8-4787-a0cf-ec363226c552)

It is ready for review again. If you have some time, could you check it?

mitsudome-r replied:
I will run them on my machine as well. Once confirmed, I think we can merge this PR.

mitsudome-r commented:
It seems my pointcloud_container keeps crashing when I launch planning_control mode. I will take a deeper look and see what's causing the error.

brkay54 commented:
> It seems my pointcloud_container keeps crashing when I launch planning_control mode. I will take a deeper look and see what's causing the error.
Thank you @mitsudome-r -san, I am going to check this.

brkay54 commented:
@mitsudome-r -san, actually I saw this issue before but I didn't focus on that issue because it does not prevent running reaction_analyzer in planning control mode. Because of it, I didn't focus on that error. Probably the error you encounter is similar.

Here log of the error:
```bash
[component_container_mt-1] [WARN 1715586401.197737367] [probabilistic_occupancy_grid_map]: Failed to lookup transform: Lookup would require extrapolation into the future.  Requested time 1715586400.691542 but the latest data is at time 1715586400.655299, when looking up transform from frame [base_link] to frame [map] (transformPointcloud() at /home/berkay/projects/autoware/src/universe/autoware.universe/perception/probabilistic_occupancy_grid_map/src/utils/utils.cpp:35)
[component_container_mt-1] terminate called after throwing an instance of 'std::runtime_error'
[component_container_mt-1]   what():  Field x does not exist
[component_container_mt-1] *** Aborted at 1715586401 (unix time) try "date -d @1715586401" if you are using GNU date ***
[component_container_mt-1] PC: @                0x0 (unknown)
[component_container_mt-1] *** SIGABRT (@0x3e8001bf6b6) received by PID 1832630 (TID 0x7f54e1ffb640) from PID 1832630; stack trace: ***
[component_container_mt-1]     @     0x7f550c40a006 google::(anonymous namespace)::FailureSignalHandler()
[component_container_mt-1]     @     0x7f5510842520 (unknown)
[component_container_mt-1]     @     0x7f55108969fc pthread_kill
[component_container_mt-1]     @     0x7f5510842476 raise
[component_container_mt-1]     @     0x7f55108287f3 abort
[component_container_mt-1]     @     0x7f5510ca2b9e (unknown)
[component_container_mt-1]     @     0x7f5510cae20c (unknown)
[component_container_mt-1]     @     0x7f5510cae277 std::terminate()
[component_container_mt-1]     @     0x7f5510cae4d8 __cxa_throw
[component_container_mt-1]     @     0x7f550470745b sensor_msgs::impl::PointCloud2IteratorBase<>::set_field()
[component_container_mt-1]     @     0x7f550470682b costmap_2d::OccupancyGridMap::raytraceFreespace()
[component_container_mt-1]     @     0x7f5504706d0b costmap_2d::OccupancyGridMap::raytrace2D()
[component_container_mt-1]     @     0x7f55046c19a6 occupancy_grid_map::LaserscanBasedOccupancyGridMapNode::onLaserscanPointCloud2WithObstacleAndRaw()
[component_container_mt-1]     @     0x7f55046ff00b message_filters::CallbackHelper9T<>::call()
[component_container_mt-1]     @     0x7f5504700da0 message_filters::sync_policies::ExactTime<>::checkTuple()
[component_container_mt-1]     @     0x7f550470267d message_filters::Synchronizer<>::cb<>()
[component_container_mt-1]     @     0x7f55046fdea8 message_filters::CallbackHelper1T<>::call()
[component_container_mt-1]     @     0x7f55046fe007 _ZNSt17_Function_handlerIFvSt10shared_ptrIKN11sensor_msgs3msg10LaserScan_ISaIvEEEEEZN15message_filters10SubscriberIS5_N6rclcpp4NodeEE9subscribeEPSC_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE17rmw_qos_profile_sNSB_32SubscriptionOptionsWithAllocatorIS4_EEEUlS7_E_E9_M_invokeERKSt9_Any_dataOS7_
[component_container_mt-1]     @     0x7f55046d6752 _ZNSt8__detail9__variant17__gen_vtable_implINS0_12_Multi_arrayIPFNS0_21__deduce_visit_resultIvEEOZN6rclcpp23AnySubscriptionCallbackIN11sensor_msgs3msg10LaserScan_ISaIvEEESA_E8dispatchESt10shared_ptrISB_ERKNS5_11MessageInfoEEUlOT_E_RSt7variantIJSt8functionIFvRKSB_EESN_IFvSP_SH_EESN_IFvRKNS5_17SerializedMessageEEESN_IFvSW_SH_EESN_IFvSt10unique_ptrISB_St14default_deleteISB_EEEESN_IFvS14_SH_EESN_IFvS11_ISU_S12_ISU_EEEESN_IFvS1A_SH_EESN_IFvSD_ISO_EEESN_IFvS1F_SH_EESN_IFvSD_ISV_EEESN_IFvS1K_SH_EESN_IFvRKS1F_EESN_IFvS1Q_SH_EESN_IFvRKS1K_EESN_IFvS1W_SH_EESN_IFvSE_EESN_IFvSE_SH_EESN_IFvSD_ISU_EEESN_IFvS25_SH_EEEEEJEEESt16integer_sequenceImJLm8EEEE14__visit_invokeESL_S2B_
[component_container_mt-1]     @     0x7f55046fb3d3 rclcpp::Subscription<>::handle_message()
[component_container_mt-1]     @     0x7f55110237bc rclcpp::Executor::execute_subscription()
[component_container_mt-1]     @     0x7f5511023fbf rclcpp::Executor::execute_any_executable()
[component_container_mt-1]     @     0x7f551102b27a rclcpp::executors::MultiThreadedExecutor::run()
[component_container_mt-1]     @     0x7f5510cdc253 (unknown)
[component_container_mt-1]     @     0x7f5510894ac3 (unknown)
[component_container_mt-1]     @     0x7f5510926850 (unknown)
[ERROR] [component_container_mt-1]: process has died [pid 1832630, exit code -6, cmd '/opt/ros/humble/lib/rclcpp_components/component_container_mt --ros-args -r __node:=pointcloud_container -r __ns:=/ -p use_sim_time:=False -p wheel_radius:=0.383 -p wheel_width:=0.235 -p wheel_base:=2.79 -p wheel_tread:=1.64 -p front_overhang:=1.0 -p rear_overhang:=1.1 -p left_overhang:=0.128 -p right_overhang:=0.128 -p vehicle_height:=2.5 -p max_steer_angle:=0.7'].
```
Let's discuss this at tomorrow's meeting.

brkay54 commented:
@mitsudome-r -san, I guess, I found the problem. When we start to publish the empty pointcloud before the initialization, it causes the crash above. We should publish the pointcloud message after the vehicle is initialized. I tried 7 times in planning_control mode and I didn't get the error we mentioned above. When you have time, could you test it again with the latest commit?

beyzanurkaya commented:
Very nice work @brkay54
I tried this PR on my own computer and I think a few minor improvements can be made.

- Firstly, when running the planning_control mode without making any changes to the param file, including file paths, I noticed that even with invalid file paths in the param file, the node worked without giving any error or warning. You can throw an error when an invalid file path is given to make sure the user can see the results after testing.

- Secondly, in the scenario created for the planning_control mode, the vehicle starts driving and stops for an object close to the goal point, abruptly ending the test. However, since the object is far from the ego's starting point, it may not be immediately clear whether there is an object at that point until the vehicle moves forward for a while. This could lead to confusion about why the test ends at that specific point. To improve clarity, you can add a marker symbolizing the object there or start the next test a few seconds after the vehicle has stopped completely.

- When I ran the perception_planning mode (with no changes made to the param file other than the bag and output file paths), I encountered an issue where the test did not complete. Although the vehicle was initialized, the goal was set, and the route was generated with objects properly detected, the test did not proceed beyond this point. No output file was created. However, when I commented out the control topics in the yaml file and retried, the tests were completed successfully, and the result file was created. In this regard, it would be useful for users to specify the reaction topics that the mods expect separately or to add a warning about this to the readme.

Also, I did not encounter any node crashes like in the last comment. I think the README file was quite descriptive. Overall, it's a well-done piece of work!

brkay54 commented:
@beyzanurkaya Thank you for the review.

> Firstly, when running the planning_control mode without making any changes to the param file, including file paths, I noticed that even with invalid file paths in the param file, the node worked without giving any error or warning. You can throw an error when an invalid file path is given to make sure the user can see the results after testing.

I have added a condition for checking whether the output path is valid or not at the beginning of the node.

> Secondly, in the scenario created for the planning_control mode, the vehicle starts driving and stops for an object close to the goal point, abruptly ending the test. However, since the object is far from the ego's starting point, it may not be immediately clear whether there is an object at that point until the vehicle moves forward for a while. This could lead to confusion about why the test ends at that specific point. To improve clarity, you can add a marker symbolizing the object there or start the next test a few seconds after the vehicle has stopped completely.

Thank you for the suggestion. I added a debug marker for the entity that will be spawned.

> When I ran the perception_planning mode (with no changes made to the param file other than the bag and output file paths), I encountered an issue where the test did not complete. Although the vehicle was initialized, the goal was set, and the route was generated with objects properly detected, the test did not proceed beyond this point. No output file was created. However, when I commented out the control topics in the yaml file and retried, the tests were completed successfully, and the result file was created. In this regard, it would be useful for users to specify the reaction topics that the mods expect separately or to add a warning about this to the readme.

Because the vehicle is not moving in perception_planning mode, the control topics may not react. Thus the reaction_analyzer waits for them to react. I added this warning in the readme, I also added a warning message prints reaction_analyzer is waiting for which node.

I appreciate your review. Could you also try with the last commit?

mitsudome-r reviewed (approved):
LGTM

beyzanurkaya commented:
LGTM

beyzanurkaya reviewed (approved):
LGTM

sribrundav commented:
running reaction analyzer but the vehicle is not initializing.. I was able to initialize the vehicle by giving a current pose and goal pose through rviz but the dynamic obstacle is still not spawning in the map. (using the default nishishinjuku map). Is this an ongoing issue or is there a work around for this ?
